### DISCUSSION TITLE
How to MapReduce ?

### DISCUSSION THREAD

**Discussion by jc3ds** [microsoft/semantic-kernel#668] [Q&A]
To follow up on issue #508, it would be great to have the capability to do some kind of MapReduce on arrays of strings.

Scenarios like WebSearch returning more than one result would benefit from
- mapping each result to a function (e.g. for summarizing, checking, re-ranking, ...)
- reducing the outputs to yield a single prompt

According to the documentation: "Branching features such as "if", "for", and code blocks are not part of SK's template language. This reflects SK's design principle of using natural language as much as possible, with a clear separation from conventional programming code."

So, what would be the recommended way to implement such a scenario?

**Comment by evchaki** [comment]
@jc3ds for a small amount of data you can use function chaining to map and filter data.  For larger scenarios we recommend using a native function for processing and then result the filtered results.

**Reply by a-bormark** [reply]
What do you mean by "native function"?

**Comment by Rehu199** [comment]
> To follow up on issue , it would be great to have the capability to do some kind of MapReduce on arrays of strings.
> 
> Scenarios like WebSearch returning more than one result would benefit from
> - mapping each result to a function (e.g. for summarizing, checking, re-ranking, ...)
> - reducing the outputs to yield a single prompt
> 
> According to the documentation: "Branching features such as "if", "for", and code blocks are not part of SK's template language. This reflects SK's design principle of using natural language as much as possible, with a clear separation from conventional programming code."
> 
> So, what would be the recommended way to implement such a scenario?

**Comment by anthonypuppo** [comment]
The extension method `AggregatePartitionedResultsAsync` is effectively a map reduce. It will apply the same function to a partitioned data set (in this case, an array of strings). You can then further "reduce" if need be or just display the aggregated result.

https://github.com/microsoft/semantic-kernel/blob/ef4c5a53a5f1e9c3bd7821b6631763082123cfc1/dotnet/src/SemanticKernel.Core/Functions/SKFunctionTextExtensions.cs#L29
